File: Leetcode/20.java
class Solution {

    static char[] openBrackets = {'(', '{', '['};

    public boolean isValid(String s) {
        Stack<Character> st = new Stack<>();
        
        for (int i=0; i<s.length(); i++) {
            char current = s.charAt(i);
            if (isOpenBracket(current)) {
                st.push(current);
                continue;
            }

            if (st.isEmpty()) return false;
            
            char savedBracket = st.pop();
            if (!match(savedBracket, current)) return false;
        }

        if (!st.isEmpty()) return false;

        return true;
    }

    boolean isOpenBracket(char c) {
        for (char b : openBrackets) {
            if (c == b) return true;
        }
        return false;
    }

    boolean match(char a, char b) {
        if (a == '(' && b == ')') return true;
        if (a == '{' && b == '}') return true;
        if (a == '[' && b == ']') return true;
        return false;
    }
}

File: Leetcode/20.cpp
class Solution {
    public:
        char counter(char c) {
            if (c == ')') return '(';
            if (c == ']') return '[';
            return '{';
        }
    
        bool isValid(string s) {
            stack<char> st;
            
            for (int i=0; i<s.size(); i++) {
                if (s[i] == ')' || s[i] == ']' || s[i] == '}') {
                    if (st.empty()) return false;
                    if (st.top() != counter(s[i])) return false;
                    st.pop();
                } else {
                    st.push(s[i]);
                }
            }
    
            if (!st.empty()) return false;
    
            return true;
        }
    };